### ISSUE TITLE
Assert error in receive_frame(), vtc_http2.c line 839: Condition((vsb) == 0) not true.

### ISSUE THREAD

**Issue by Dridi** [varnishcache/varnish-cache#2256]
Seen on Travis CI (Ubuntu, Clang):

```
FAIL: tests/t02007
==================
**** top   0.0 extmacro def pwd=/home/travis/build/varnishcache/varnish-cache/bin/varnishtest
**** top   0.0 extmacro def localhost=127.0.0.1
**** top   0.0 extmacro def bad_backend=127.0.0.1 46745
**** top   0.0 extmacro def bad_ip=192.0.2.255
**** top   0.0 extmacro def topbuild=/home/travis/build/varnishcache/varnish-cache
**** top   0.0 macro def tmpdir=/tmp/vtc.49576.30e5d90d
*    top   0.0 TEST ./tests/t02007.vtc starting
**   top   0.0 === varnishtest "H2 Huge response headers"
*    top   0.0 TEST H2 Huge response headers
**   top   0.0 === server s1 {
**   s1    0.0 Starting server
**** s1    0.0 macro def s1_addr=127.0.0.1
**** s1    0.0 macro def s1_port=41579
**** s1    0.0 macro def s1_sock=127.0.0.1 41579
*    s1    0.0 Listen on 127.0.0.1 41579
**   top   0.0 === varnish v1 -vcl+backend {} -cliok "param.set feature +http2"...
**   s1    0.0 Started on 127.0.0.1 41579
**   v1    0.0 Launch
***  v1    0.0 CMD: cd ${pwd} && exec varnishd  -d -n /tmp/vtc.49576.30e5d90d/v1 -l 2m,1m,- -p auto_restart=off -p syslog_cli_traffic=off -p sigsegv_handler=on -p thread_pool_min=10 -p debug=+vtc_mode -a '127.0.0.1:0' -M '127.0.0.1 32899' -P /tmp/vtc.49576.30e5d90d/v1/varnishd.pid -p vmod_path=/home/travis/build/varnishcache/varnish-cache/lib/libvmod_std/.libs:/home/travis/build/varnishcache/varnish-cache/lib/libvmod_debug/.libs:/home/travis/build/varnishcache/varnish-cache/lib/libvmod_directors/.libs 
***  v1    0.0 CMD: cd /home/travis/build/varnishcache/varnish-cache/bin/varnishtest && exec varnishd  -d -n /tmp/vtc.49576.30e5d90d/v1 -l 2m,1m,- -p auto_restart=off -p syslog_cli_traffic=off -p sigsegv_handler=on -p thread_pool_min=10 -p debug=+vtc_mode -a '127.0.0.1:0' -M '127.0.0.1 32899' -P /tmp/vtc.49576.30e5d90d/v1/varnishd.pid -p vmod_path=/home/travis/build/varnishcache/varnish-cache/lib/libvmod_std/.libs:/home/travis/build/varnishcache/varnish-cache/lib/libvmod_debug/.libs:/home/travis/build/varnishcache/varnish-cache/lib/libvmod_directors/.libs 
***  v1    0.0 PID: 49595
**** v1    0.0 macro def v1_pid=49595
**** v1    0.0 macro def v1_name=/tmp/vtc.49576.30e5d90d/v1
***  v1    0.0 debug|Debug: Platform: Linux,4.8.12-040812-generic,x86_64,-jnone,-smalloc,-smalloc,-hcritbit
***  v1    0.0 debug|200 287     
***  v1    0.0 debug|-----------------------------
***  v1    0.0 debug|Varnish Cache CLI 1.0
***  v1    0.0 debug|-----------------------------
***  v1    0.0 debug|Linux,4.8.12-040812-generic,x86_64,-jnone,-smalloc,-smalloc,-hcritbit
***  v1    0.0 debug|varnish-trunk revision 7c328f9
***  v1    0.0 debug|
***  v1    0.0 debug|Type 'help' for command list.
***  v1    0.0 debug|Type 'quit' to close CLI session.
***  v1    0.0 debug|Type 'start' to launch worker process.
***  v1    0.0 debug|
**** v1    0.1 CLIPOLL 1 0x1 0x0
***  v1    0.1 CLI connection fd = 10
***  v1    0.1 CLI RX  107
**** v1    0.1 CLI RX|ddcuwajfhmjvqwfiwkwjtqcuphxvmzwo
**** v1    0.1 CLI RX|
**** v1    0.1 CLI RX|Authentication required.
**** v1    0.1 CLI TX|auth 96dd9c29538c67e9ead148bbe8a82e7328709b7e3b08a622d9b154c704c864a1
***  v1    0.1 CLI RX  200
**** v1    0.1 CLI RX|-----------------------------
**** v1    0.1 CLI RX|Varnish Cache CLI 1.0
**** v1    0.1 CLI RX|-----------------------------
**** v1    0.1 CLI RX|Linux,4.8.12-040812-generic,x86_64,-jnone,-smalloc,-smalloc,-hcritbit
**** v1    0.1 CLI RX|varnish-trunk revision 7c328f9
**** v1    0.1 CLI RX|
**** v1    0.1 CLI RX|Type 'help' for command list.
**** v1    0.1 CLI RX|Type 'quit' to close CLI session.
**** v1    0.1 CLI RX|Type 'start' to launch worker process.
**** v1    0.1 CLI TX|vcl.inline vcl1 << %XJEIFLH|)Xspa8P
**** v1    0.1 CLI TX|vcl 4.0;
**** v1    0.1 CLI TX|backend s1 { .host = "127.0.0.1"; .port = "41579"; }
**** v1    0.1 CLI TX|
**** v1    0.1 CLI TX|
**** v1    0.1 CLI TX|%XJEIFLH|)Xspa8P
***  v1    0.2 CLI RX  200
**** v1    0.2 CLI RX|VCL compiled.
**** v1    0.2 CLI TX|vcl.use vcl1
***  v1    0.2 CLI RX  200
**** v1    0.2 CLI TX|param.set feature +http2
***  v1    0.3 CLI RX  200
**   v1    0.3 CLI 200 <param.set feature +http2>
**   v1    0.3 Start
**** v1    0.3 CLI TX|start
***  v1    0.3 debug|Debug: Child (49640) Started
***  v1    0.4 CLI RX  200
***  v1    0.4 wait-running
**** v1    0.4 CLI TX|status
***  v1    0.4 debug|Info: Child (49640) said Child starts
***  v1    0.4 CLI RX  200
**** v1    0.4 CLI RX|Child in state running
**** v1    0.4 CLI TX|debug.xid 999
**** v1    0.4 vsl|          0 CLI             - Rd vcl.load "vcl1" vcl_vcl1.1489409616.776020527/vgc.so 1auto
**** v1    0.4 vsl|          0 CLI             - Wr 200 55 Loaded "vcl_vcl1.1489409616.776020527/vgc.so" as "vcl1"
**** v1    0.4 vsl|          0 CLI             - Rd vcl.use "vcl1"
**** v1    0.4 vsl|          0 CLI             - Wr 200 0 
**** v1    0.4 vsl|          0 CLI             - Rd start
**** v1    0.4 vsl|          0 CLI             - Wr 200 0 
***  v1    0.4 CLI RX  200
**** v1    0.4 CLI RX|XID is 999
**** v1    0.4 CLI TX|debug.listen_address
***  v1    0.5 CLI RX  200
**** v1    0.5 CLI RX|127.0.0.1 46393
**   v1    0.5 Listen on 127.0.0.1 46393
**** v1    0.5 macro def v1_addr=127.0.0.1
**** v1    0.5 macro def v1_port=46393
**** v1    0.5 macro def v1_sock=127.0.0.1 46393
**   top   0.5 === varnish v1 -cliok "param.set debug +syncvsl"
**** v1    0.5 CLI TX|param.set debug +syncvsl
**** v1    0.5 vsl|          0 CLI             - Rd debug.xid 999 
**** v1    0.5 vsl|          0 CLI             - Wr 200 10 XID is 999
**** v1    0.5 vsl|          0 CLI             - Rd debug.listen_address 
**** v1    0.5 vsl|          0 CLI             - Wr 200 16 127.0.0.1 46393
***  v1    0.5 CLI RX  200
**   v1    0.5 CLI 200 <param.set debug +syncvsl>
**   top   0.5 === varnish v1 -cliok "param.set debug +h2_nocheck"
**** v1    0.5 CLI TX|param.set debug +h2_nocheck
***  v1    0.6 CLI RX  200
**   v1    0.6 CLI 200 <param.set debug +h2_nocheck>
**   top   0.6 === client c1 {
**   c1    0.6 Starting client
**   c1    0.6 Waiting for client
***  c1    0.6 Connect to 127.0.0.1 46393
***  c1    0.6 connected fd 11 from 127.0.0.1 40398 to 127.0.0.1 46393
**   c1    0.6 === stream 0 {
**** c1    0.6 Not in H/2 mode, do what's needed
**   c1    0.6 === txpri
**** c1    0.6 txpri|PRI * HTTP/2.0\r
**** c1    0.6 txpri|\r
**** c1    0.6 txpri|SM\r
**** c1    0.6 txpri|\r
**   c1    0.6 === stream 0 {
**   c1    0.6 Starting stream 0x2b41ac001350
**   c1    0.6 Waiting for stream 0
**   c1    0.6 === txsettings
***  c1    0.6 tx: stream: 0, type: SETTINGS (4), flags: 0x00, size: 0
**   c1    0.6 === rxsettings
***  c1    0.6 rx: stream: 0, type: SETTINGS (4), flags: 0x00, size: 12
**** c1    0.6 settings->MAX_CONCURRENT_STREAMS (3): 100
**** c1    0.6 settings->INITIAL_WINDOW_SIZE (4): 65535
**   c1    0.6 === txsettings -ack
***  c1    0.6 tx: stream: 0, type: SETTINGS (4), flags: 0x01, size: 0
**   c1    0.6 === rxsettings
***  c1    0.6 rx: stream: 0, type: SETTINGS (4), flags: 0x01, size: 0
***  c1    0.6 flag: ACK
**   c1    0.6 === expect settings.ack == true
**** c1    0.6 (s0) EXPECT settings.ack (true) == "true" match
**   c1    0.6 Ending stream 0
**   c1    0.6 Starting stream 0x2b41ac001350
**   c1    0.6 Waiting for stream 0
**   c1    0.6 === txsettings -framesize 64
***  c1    0.6 tx: stream: 0, type: SETTINGS (4), flags: 0x00, size: 6
**   c1    0.6 === rxsettings
***  c1    0.6 rx: stream: 0, type: SETTINGS (4), flags: 0x01, size: 0
***  c1    0.6 flag: ACK
**   c1    0.6 Ending stream 0
**   c1    0.6 === stream 1 {
**   c1    0.6 Starting stream 0x2b41ac002590
**   c1    0.6 Waiting for stream 1
**   c1    0.6 === txreq \
***  c1    0.6 tx: stream: 1, type: HEADERS (1), flags: 0x00, size: 62
**   c1    0.6 === txcont \
***  c1    0.6 tx: stream: 1, type: CONTINUATION (9), flags: 0x04, size: 39
**   c1    0.6 === txdata \
***  c1    0.6 tx: stream: 1, type: DATA (0), flags: 0x01, size: 7
**   c1    0.6 === rxhdrs
***  c1    0.6 rx: stream: 1, type: HEADERS (1), flags: 0x04, size: 5
***  c1    0.6 flag: END_TYPE_HEADERS
**** c1    0.6 header[ 0]: :status : 100
**   c1    0.6 === expect resp.status == 100
***  s1    0.6 accepted fd 5
**** c1    0.6 (s1) EXPECT resp.status (100) == "100" match
**   s1    0.6 === rxreq
**   c1    0.6 === rxwinup
**** s1    0.6 rxhdr|POST / HTTP/1.1\r
**** s1    0.6 rxhdr|content-type: text/plain\r
**** s1    0.6 rxhdr|scheme: http\r
**** s1    0.6 rxhdr|content-length: 7\r
**** s1    0.6 rxhdr|X-Forwarded-For: 127.0.0.1\r
**** s1    0.6 rxhdr|X-Varnish: 1002\r
**** s1    0.6 rxhdr|Host: 127.0.0.1\r
**** s1    0.6 rxhdr|Transfer-Encoding: chunked\r
**** s1    0.6 rxhdr|\r
**** s1    0.6 rxhdrlen = 168
**** s1    0.6 http[ 0] |POST
**** s1    0.6 http[ 1] |/
**** s1    0.6 http[ 2] |HTTP/1.1
**** s1    0.6 http[ 3] |content-type: text/plain
**** s1    0.6 http[ 4] |scheme: http
**** s1    0.6 http[ 5] |content-length: 7
**** s1    0.6 http[ 6] |X-Forwarded-For: 127.0.0.1
**** s1    0.6 http[ 7] |X-Varnish: 1002
**** s1    0.6 http[ 8] |Host: 127.0.0.1
**** s1    0.6 http[ 9] |Transfer-Encoding: chunked
**** s1    0.6 len|007\r
**** s1    0.6 chunk|request
**** s1    0.6 len|0\r
**** s1    0.6 bodylen = 7
**   s1    0.6 === expect req.proto == HTTP/1.1
**** s1    0.6 EXPECT req.proto (HTTP/1.1) == "HTTP/1.1" match
**   s1    0.6 === txresp -hdr "Content-Type: text/plain" -hdrlen Foo 100 -body...
**** s1    0.6 txresp|HTTP/1.1 200 OK\r
**** s1    0.6 txresp|Content-Type: text/plain\r
**** s1    0.6 txresp|Foo: 9876543210987654321098765432109876543210987654321098765432109876543210987654321098765432109876543210\r
**** s1    0.6 txresp|Content-Length: 100\r
**** s1    0.6 txresp|\r
**** s1    0.6 txresp|!"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_
**** s1    0.6 txresp|"#$%&'()*+,-./0123456789:;<=>?@ABCD
***  s1    0.6 shutting fd 5
**   s1    0.6 Ending
***  c1    0.6 rx: stream: 1, type: HEADERS (1), flags: 0x00, size: 64
**   c1    0.6 === rxresp
***  c1    0.6 rx: stream: 1, type: CONTINUATION (9), flags: 0x00, size: 64
---- c1    0.6 Frame #1 for rxresp was of type CONTINUATION (9) instead of HEADERS (1)
*    diag  0.0 Assert error in receive_frame(), vtc_http2.c line 839:
*    diag  0.0   Condition((vsb) == 0) not true.
#    top  TEST ./tests/t02007.vtc FAILED (0.749) signal=6
```

**Comment by bsdphk** [comment]
This is a problem in varnishtest which causes random H2 failures.

Look near the tail of the trace, and you see c1 getting and appearantly swallowing the HEADERS right before the "rxresp" command which should handle it is executed.

**Comment by Dridi** [comment]
I know, I can have a look later but before I restarted the Travis job I wanted to log the error on github. You can assign this one to me, nothing urgent IMO.

**Comment by Dridi** [comment]
@fgsch I didn't put the `c=H/2` label (I just noticed you did) because I'm assuming it means H2 in `varnishd` (looking at other tickets with this label). It was also to avoid bothering @bsdphk while he's squashing H2 stuff before the release.

**Comment by fgsch** [comment]
We have a separate label for varnishd and this ticket is a good example why
we have the distinction IMO.

You can search for specific labels, e.g. H/2 and varnishd if you want to.

On 14 Mar 2017 10:07 am, "Dridi Boukelmoune" <notifications@github.com>
wrote:

@fgsch <https://github.com/fgsch> I didn't put the c=H/2 label (I just
noticed you did) because I'm assuming it means H2 in varnishd (looking at
other tickets with this label). It was also to avoid bothering @bsdphk
<https://github.com/bsdphk> while he's squashing H2 stuff before the
release.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/varnishcache/varnish-cache/issues/2256#issuecomment-286376103>,
or mute the thread

**Comment by Dridi** [comment]
Yes, that's why I didn't remove it.

**Comment by Dridi** [comment]
Fixed in d87dcf1

**Comment by hermunn** [comment]
Backport review: H2 bug, not in 4.1
